src/Nethermind/Nethermind.Serialization.FluentRlp.Generator/RlpSourceGenerator.cs

```diff
@@ -0,0 +1,320 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using System.Text;
+
+namespace Nethermind.Serialization.FluentRlp.Generator;
+
+public enum RlpRepresentation : byte
+{
+    /// <summary>
+    /// The RLP encoding will be a sequence of RLP objects for each property.
+    /// </summary>
+    Record = 0,
+
+    /// <summary>
+    /// The RLP encoding will be equivalent to the only underlying property.
+    /// </summary>
+    Newtype = 1,
+}
+
+[AttributeUsage(AttributeTargets.Class)]
+public sealed class RlpSerializable(RlpRepresentation representation = RlpRepresentation.Record) : Attribute
+{
+    public RlpRepresentation Representation { get; } = representation;
+}
+
+/// <summary>
+/// A source generator that finds all records with [RlpSerializable] attribute and
+/// generates an abstract `IRlpConverter` class with `Read` and `Write` methods.
+/// </summary>
+[Generator]
+public sealed class RlpSourceGenerator : IIncrementalGenerator
+{
+    private const string Version = "0.1";
+    private const string GeneratedCodeAttribute = $"""[GeneratedCode("{nameof(RlpSourceGenerator)}", "{Version}")]""";
+
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var provider = context.SyntaxProvider.CreateSyntaxProvider(
```

Comment:
Consider attribute based creation using [ForAttributeWithMetadataName](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.syntaxvalueprovider.forattributewithmetadataname?view=roslyn-dotnet-4.9.0). It should be much more cheaper than scanning all the records and only then select on attribute basis.